### ISSUE TITLE
"Form control label is descriptive" (cc0f0a) is broken…

### ISSUE THREAD

**Issue by Jym77** [act-rules/act-rules.github.io#953]
So, during the ACT-R face to face meeting, @audreymaniez  figured out that the update of rule cc0f0a (now “Form control label is descriptive”) was deeply flawed in that one of the rules from the split does not have the fail-on-fail relationship required. See https://github.com/act-rules/act-rules.github.io/pull/905#discussion_r328074530

---
While looking at the SC and trying to figure out what to do, I realised that the [**current** rule](https://act-rules.github.io/rules/cc0f0a) is already broken (I think).
If we consider (almost) the same example:
```html
<div>Name</div>
<label>Fill in: <input type=”text”/></label>
```
This example fails cc0f0a (it has a `label` element which does not describe the purpose of the associated field). However, this does not fails [SC 2.4.6](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html) since the `div` element describes the purpose of the field and is a ["WCAG label"](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-label) (text that is presented to identify a component).
(I’ll keep writing "WCAG label" in double quotes to refer to the WCAG sense of the term)

Note that this does fails [SC 1.3.1 (Info and Relationship)](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships) since the "WCAG label" (`div` element) is not programmatically associated with the form field, but this is irrelevant for SC 2.4.6.
Hence, the **current** rule does not have a fail-on-fail relationship with the SC and this is **bad**.

---
The update work on this rule, that was started by @annethyme, was considering `label` elements and accessible names separately. Reading the SC, I am not even sure why. The definition of "WCAG labels" has a note saying:
> A label is presented to all users whereas the name may be hidden and only exposed by assistive technology. 

Together with
> This success criteria does not require the use of labels; however, it does require that if labels are present, they must be sufficiently clear or descriptive. 

it looks like having an accessible name but no "WCAG label" makes the SC irrelevant (not failing, whatever the accessible name is). Even through that is likely to be a failure for other SC. In other words, an example like
```html
<input type="text" aria-label="fill in"/>
```
should not fail SC 2.4.6 because there is no "WCAG label" and the SC does not require them but require that **if they are present**, then they are descriptive.

I was not here when the rule was written nor when the rewrite started, so I have little clues on the discussions that went on at that point. The original PR (#447) was made to close issue #423 but that looks unrelated to accessible names.

---
What to do?

We clearly need to do something about the current rule (if my analysis that it is broken is true…) 

My understanding of SC 2.4.6 makes me believe that we cannot treat `label` elements independently from "WCAG labels", or rather than anything concerned only about `label` elements will not be enough to fail the SC since there could always be a "WCAG label" around fixing the problem.

There could, of course, be an atomic rule concerned about `label` elements and without accessibility requirement to SC 2.4.6 (essentially, it would be the [current proposal](https://github.com/act-rules/act-rules.github.io/blob/7ec37f139a363665c75418aa767b28443fb9bf02/_rules/form-field-visual-label-is-descriptive-1cd4da.md)). But it would be of little use alone, and we need to handle "WCAG labels" if we want to make real progress on that.

The not too bad part is that the latest PR (#905) already has some work on [context](https://github.com/act-rules/act-rules.github.io/blob/7ec37f139a363665c75418aa767b28443fb9bf02/pages/glossary/context.md) and notably [visual context](https://github.com/act-rules/act-rules.github.io/blob/7ec37f139a363665c75418aa767b28443fb9bf02/pages/glossary/visual-context.md) which can be a good basis to have a definition of "WCAG label" (we need another name) that could be used to fix the rule.

A quick fix might be to change the rule to "`label` element in context is descriptive" which would pass the example above and be OK. But essentially what we want is "context of form control is descriptive" so it is a bit of a waste to artificially focus on `label` elements when the actual "good" rule (if we ever write it) would supersede that…

At this point, I need inputs and discussions from people with better understanding of accessibility issues and WCAG than I do…

**Comment by Jym77** [comment]
Possible solution: add assumption that the `label` is the only "WCAG label".

**Comment by Jym77** [comment]
Thisis actually not a very good solution. The full update from that rule started with #423 which is taking the context into account and assuming that `label` element is all that we have goes against that.

**Comment by ShadowBB** [comment]
Wilco and myself talked about this issue extensively today.

We came to the conclusion that 2.4.6 does not require a descriptive accessible name, instead 2.4.6 is limited to visible labels only.

This means that the only two things that should be changed in the current rule are:

- "Included in the accessibility tree" should be removed from the applicability.
- The following assumption needs to be added:
> The rule assumes that the 'label' element contains the complete label (link to WCAG definition of label).

From reading the understanding documents in became clear to use that the accessibile name is not part of the label.
> From https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html
> Further, this Success Criterion does not take into consideration whether or not alternative methods of providing an accessible name or description for form controls and inputs has been used - this aspect is covered separately by 4.1.2: Name, Role and Value. It is possible for controls and inputs to have an appropriate accessible name or description (e.g. using aria-label="...") and therefore pass Success Criterion 4.1.2, but to still fail this Success Criterion (if the labels or instructions aren't presented to all users, not just those using assistive technologies).

**Comment by Jym77** [comment]
> * The following assumption needs to be added:
> 
> > The rule assumes that the 'label' element contains the complete label (link to WCAG definition of label).

@ShadowBB I agree that this would fix the current rule.

However,
1. the rewrite of #447/#905 started with #423 which is all about adding context (at least to differentiate between identical `label`s). I find it very weird to assume that everything is in the `label` and at the same time look around it. Does that mean we're completely ignoring #423?
2. I think that we can change `label` by "WCAG label" in the rule and get it OK without the assumption (and including context). That would become more manual but also cover strictly more stuff.

**Comment by Jym77** [comment]
Should be fixed by #973.
